File: Data_Structure/DoublyLinearLL.java
class node
{
    public int data;
    public node next;
    public node prev;

    public node(int iNo)
    {
        this.data = iNo;
        this.next = null;
        this.prev = null;
    }
}

class DoublyLLL
{
    private node first;
    private int iCount;

    public DoublyLLL()
    {
        this.first = null;
        this.iCount = 0;
    }

    public void InsertFirst(int iNo)
    {
        node newn = null;

        newn = new node(iNo);

        if(null == first) 
        {
            first = newn;
        }
        else
        {
            newn.next = first;
            first.prev = newn;
            first = newn;
        }
        iCount++;
    }

    public void InsertLast(int iNo)
    {
        node newn = null, temp = null;

        newn = new node(iNo);

        if(null == first) 
        {
            first = newn;
        }
        else
        {
            temp = first;

            while(temp.next != null) 
            {
                temp = temp.next;
            }
            temp.next = newn;
            newn.prev = temp;
        }
        iCount++;
    }

    public void InsertAtPos(int iNo, int iPos)
    {
        node newn = null, temp = null;
        int iCnt = 0;

        if((iPos < 1) || (iPos > iCount+1)) 
        {
            System.out.print("Invalid Position");
            return;
        }

        if(iPos == 1) 
        {
            InsertFirst(iNo);
        }
        else if(iPos == iCount+1) 
        {
            InsertLast(iNo);
        }
        else
        {
            newn = new node(iNo);

            temp = first;

            for(iCnt = 1;iCnt < iPos-1;iCnt++) 
            {
                temp = temp.next;
            }

            newn.next = temp.next;
            temp.next.prev = newn;
            temp.next = newn;
            newn.prev = temp;

            iCount++;
        }
    }

    public void DeleteFirst()
    {
        if(null == first)
        {
            return;
        }
        else if(first.next == null)
        {
            first = null;
        }
        else
        {
            first = first.next;
            first.prev = null;
        }
        iCount--;
    }

    public void DeleteLast()
    {
        node temp = null;

        if(null == first)
        {
            return;
        }
        else if(first.next == null)
        {
            first = null;
        }
        else
        {
            temp = first;

            while(temp.next.next != null) 
            {
                temp = temp.next;
            }
            temp.next = null;
        }
        iCount--;
    }

    public void DeleteAtPos(int iPos)
    {
        node temp = null;
        int iCnt = 0;

        if((iPos < 1) || (iPos > iCount)) 
        {
            System.out.print("Invalid Position");
            return;
        }

        if(iPos == 1) 
        {
            DeleteFirst();
        }
        else if(iPos == iCount) 
        {
            DeleteLast();
        }
        else
        {
            temp = first;
            for(iCnt = 1;iCnt < iPos-1;iCnt++) 
            {
                temp = temp.next;
            }
            temp.next = temp.next.next;
            temp.next.prev = temp;

            iCount--;
        }
    }

    public void Display()
    {
        node temp = null;
        temp = first;
        System.out.print("null<-");
        while(temp != null) 
        {
            System.out.print("| "+temp.data+" |<=>");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public int Count()
    {
        return iCount;
    }
}

public class DoublyLinearLL 
{
    public static void main(String A[])
    {
        DoublyLLL dobj = null;
        int iRet = 0;

        dobj = new DoublyLLL();

        dobj.InsertFirst(51);
        dobj.InsertFirst(21);
        dobj.InsertFirst(11);

        dobj.InsertLast(101);
        dobj.InsertLast(111);
        dobj.InsertLast(121);

        dobj.Display();

        iRet = dobj.Count();

        System.out.println("Number of elements in LinkedList are : "+iRet);

        dobj.DeleteFirst();

        dobj.Display();

        iRet = dobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        dobj.DeleteLast();

        dobj.Display();

        iRet = dobj.Count();

        System.out.println("Number of elements in the LinkedList are : "+iRet);

        dobj.InsertAtPos(105,5);

        dobj.Display();

        iRet = dobj.Count();

        System.out.println("Number of elements in LinkedList are : "+iRet);

        dobj.DeleteAtPos(5);

        dobj.Display();

        iRet = dobj.Count();

        System.out.println("Number of elements in LinkedList are : "+iRet);
    }    
}

File: Data_Structure/DoublyLinearLL.cpp
#include<iostream>
using namespace std;

struct node
{
    int data;
    struct node* next;
    struct node* prev;
};

typedef struct node NODE;
typedef struct node* PNODE;

class DoublyLLL
{
    private:
        PNODE first;
        int iCount;

    public:
        DoublyLLL()
        {
            this->first = NULL;
            this->iCount = 0;
        }

        void InsertFirst(int iNo)
        {
            PNODE newn = NULL;

            newn = new NODE;
            newn->data = iNo;
            newn->next = NULL;
            newn->prev = NULL;

            if(NULL == first)
            {
                first = newn;
            }
            else
            {
                newn->next = first;
                newn->next->prev = newn;
                first = newn;
            }

            iCount++;
        }
        void InsertLast(int iNo)
        {
            PNODE newn = NULL, temp = NULL;

            newn = new NODE;
            newn->data = iNo;
            newn->next = NULL;
            newn->prev = NULL;

            if(NULL == first)
            {
                first = newn;
            }
            else
            {
                temp = first;

                while(temp->next != NULL)
                {
                    temp = temp->next;
                }
                
                temp->next = newn;
                newn->prev = temp;
            }

            iCount++;
        }
        void DeleteFirst()
        {
            PNODE temp = NULL;

            if(NULL == first)
            {
                return;
            }
            else if(first->next == NULL)
            {
                delete first;
                first = NULL;
            }
            else
            {
                temp = first;
                first = first->next;
                first->prev = NULL;
                delete temp;
            }

            iCount--;
        }
        void DeleteLast()
        {
            PNODE temp = NULL;

            if(NULL == first)
            {
                return;
            }
            else if(first->next == NULL)
            {
                delete first;
                first = NULL;
            }
            else
            {
                temp = first;

                while(temp->next->next != NULL)
                {
                    temp = temp->next;
                }
                
                delete temp->next;
                temp->next = NULL;
            }

            iCount--;
        }
        void Display()
        {
            PNODE temp = NULL;
            temp = first;

            cout<<"NULL<-";
            while(temp != NULL)
            {
                cout<<"| "<<temp->data<<" |"<<"<=>";
                temp = temp->next;
            }
            cout<<"NULL\n";
        }
        int Count()
        {
            return iCount;
        }
        void InsertAtPos(int iNo, int iPos)
        {
            PNODE newn = NULL , temp = NULL;
            int iCnt = 0;

            if((iPos < 1) || (iPos > iCount+1))
            {
                cout<<"Invalid position";
                return;
            }
            
            if(iPos == 1)
            {
                InsertFirst(iNo);
            }
            else if(iPos == iCount+1)
            {
                InsertLast(iNo);
            }
            else
            {
                newn = new NODE;
                newn->data = iNo;
                newn->next = NULL;
                newn->prev = NULL;

                temp = first;

                for(iCnt = 1;iCnt < iPos-1;iCnt++)
                {
                    temp = temp->next;
                }
                
                newn->next = temp->next;
                newn->prev = temp;
                temp->next = newn;

                iCount++;
            }
        }
        void DeleteAtPos(int iPos)
        {
            PNODE temp = NULL, target = NULL;
            int iCnt = 0;

            if((iPos < 1) || (iPos > iCount))
            {
                cout<<"Invalid position";
                return;
            }
            
            if(iPos == 1)
            {
                DeleteFirst();
            }
            else if(iPos == iCount)
            {
                DeleteLast();
            }
            else
            {
                temp = first;

                for(iCnt = 1;iCnt < iPos-1;iCnt++)
                {
                    temp = temp->next;
                }
                
                target = temp->next;
                temp->next = target->next;
                target->next->prev = temp;
                delete target;

                iCount--;
            }
        }
};

int main()
{
    DoublyLLL dobj;
    int iRet = 0;

    dobj.InsertFirst(51);
    dobj.InsertFirst(21);
    dobj.InsertFirst(11);
    
    dobj.InsertLast(101);
    dobj.InsertLast(111);
    dobj.InsertLast(121);
    dobj.InsertLast(151);

    dobj.Display();

    iRet = dobj.Count();

    cout<<"Number of nodes in Linked List : "<<iRet<<"\n";

    dobj.InsertAtPos(125,7);

    dobj.Display();

    iRet = dobj.Count();

    cout<<"Number of nodes in Linked List : "<<iRet<<"\n";

    dobj.DeleteAtPos(5);

    dobj.Display();

    iRet = dobj.Count();

    cout<<"Number of nodes in Linked List : "<<iRet<<"\n";

    return 0;
}